### PR TITLE
Adding a validation method to validate the count of a value

### PR DESCRIPTION
**Pull request by burzum**
Used to check the count of a given value of type string, int, array or countable. If a string value is passed the string length is used as count.

### CODE REVIEW

File: src/Validation/Validation.php [hunk 1]
```diff
@@ -216,6 +216,35 @@ public static function cc($check, $type = 'fast', $deep = false, $regex = null)
...
+    {
+        if (is_array($check1) || $check1 instanceof \Countable) {
+            $count = count($check1);
+        } elseif (is_string($check1)) {
+            $count = mb_strlen($check1);
+        } elseif (is_int($check1)) {
+            $count = $check1;
+        }
+
+        if (!$count) {
```
Comment (markstory) [inline src/Validation/Validation.php]: What if count is 0 and you wanted to ensure the count was 0?

Comment (burzum) [inline src/Validation/Validation.php]: Fixed.

**Comment by burzum** [comment]
@dereuromark how is the null check you propose better than isset() here?

**Comment by dereuromark** [comment]
@burzum What you just did is n anti-pattern. See the discussion about wrong usage of isset and !empty() cloaking possible typos etc. Always prefer non-cloaking check if possible, and in this case it is.

**Comment by burzum** [comment]
@dereuromark 

```php
        if ($count === null) {
            return false;
        }
```

results in `Undefined variable: count` as I expected.

**Comment by dereuromark** [comment]
:+1: 
The only think I ask myself is, do people understand that Validation::count() is a validation method of "count" - or do they think this counts the validation rules or sth and returns an integer?
"isCount" as name might prevent this misleading. But it is just a question at this point.

File: tests/TestCase/Validation/ValidationTest.php [hunk 2]
```diff
@@ -2762,4 +2762,33 @@ public function testUtf8Extended()
...
+
+        $int = 7;
+        $this->assertTrue(Validation::count($int, '==', 7));
+        $this->assertFalse(Validation::count($int, '>', 8));
+        $this->assertFalse(Validation::count($int, '<', 1));
+
+        $int = 0;
+        $this->assertTrue(Validation::count($int, '==', 0));
+
+        $this->assertFalse(Validation::count(null, '==', 0));
```
Comment (markstory) [inline tests/TestCase/Validation/ValidationTest.php]: You could be extra thorough and include test cases for `true` and `false` as well.

File: src/Validation/Validation.php [hunk 3]
```diff
@@ -216,6 +216,36 @@ public static function cc($check, $type = 'fast', $deep = false, $regex = null)
...
+     * @return bool Success
+     */
+    public static function count($check1, $operator, $expectedCount)
+    {
+        $count = null;
+        if (is_array($check1) || $check1 instanceof \Countable) {
+            $count = count($check1);
+        } elseif (is_string($check1)) {
+            $count = mb_strlen($check1);
+        } elseif (is_int($check1)) {
```
Comment (lorenzo) [inline src/Validation/Validation.php]: I think this method should only array, traversable or countable stuff. We already have method for intgers and strings

File: src/Validation/Validation.php [hunk 4]
```diff
@@ -216,6 +216,36 @@ public static function cc($check, $type = 'fast', $deep = false, $regex = null)
...
+     * If a string value is passed the string length is used as count.
+     *
+     * @param array|int|string $check1 The value to check the count on.
+     * @param string $operator Can be either a word or operand
+     *    is greater >, is less <, greater or equal >=
+     *    less or equal <=, is less <, equal to ==, not equal !=
+     * @param int $expectedCount The expected count value.
+     * @return bool Success
+     */
+    public static function count($check1, $operator, $expectedCount)
```
Comment (lorenzo) [inline src/Validation/Validation.php]: Can you please rename this to `numElements()`

File: src/Validation/Validation.php (hunk 3 above)
Comment (burzum) [inline src/Validation/Validation.php]: Uh where? I think this method is pretty convenient now as it is.

Comment (lorenzo) [inline src/Validation/Validation.php]: `Validation::equals()` validates numbers, `Validation::lengthBetween()` validates string length. This method should only validate stuff that can be safely iterated.

If you think we need more / better methods for strings or integers, you are very welcome to add them.

Comment (burzum) [inline src/Validation/Validation.php]: Would you mind to remove it from the PR if you want to get that removed?

Comment (lorenzo) [inline src/Validation/Validation.php]: Sure, I can do that tomorrow after work

Comment (burzum) [inline src/Validation/Validation.php]: I strongly disagree with that but fine, I've removed it. It was just convenient.

There is no equals, just equalTo and it does not allow me to take a string and compare the count of it to something. http://api.cakephp.org/3.2/source-class-Cake.Validation.Validation.html#592 In fact it's just a strong typed comparison.

lengthBetween requirs a min and max value. http://api.cakephp.org/3.2/source-class-Cake.Validation.Validation.html#121

Both don't do the same my implementation allowed you to do.

**Comment by dereuromark** [comment]
Even better now :+1:

**Comment by burzum** [comment]
@dereuromark well I consider it a huge step back. See my comment on the change.

File: src/Validation/Validation.php [hunk 5]
```diff
@@ -216,6 +216,26 @@ public static function cc($check, $type = 'fast', $deep = false, $regex = null)
...
+     * @param array|int|string $check1 The value to check the count on.
+     * @param string $operator Can be either a word or operand
+     *    is greater >, is less <, greater or equal >=
+     *    less or equal <=, is less <, equal to ==, not equal !=
+     * @param int $expectedCount The expected count value.
+     * @return bool Success
+     */
+    public static function numElements($check1, $operator, $expectedCount)
+    {
+        if (!is_array($check1) || $check1 instanceof \Countable) {
```
Comment (chinpei215) [inline src/Validation/Validation.php]: `!` is needed before `$check1 instanceof \Countable`. Adding `ArrayObject` test might be nice.

File: src/Validation/Validation.php [hunk 6]
```diff
@@ -216,6 +216,26 @@ public static function cc($check, $type = 'fast', $deep = false, $regex = null)
...
+     * If a string value is passed the string length is used as count.
+     *
+     * @param array|int|string $check1 The value to check the count on.
+     * @param string $operator Can be either a word or operand
+     *    is greater >, is less <, greater or equal >=
+     *    less or equal <=, is less <, equal to ==, not equal !=
+     * @param int $expectedCount The expected count value.
+     * @return bool Success
+     */
+    public static function numElements($check1, $operator, $expectedCount)
```
Comment (ADmad) [inline src/Validation/Validation.php]: Keeping the var name `$check` would be nice for consistency with other methods in class.

File: src/Validation/Validation.php (hunk 3 above)
Comment (lorenzo) [inline src/Validation/Validation.php]: I don't mind adding more methods to the validation if you want this for strings

**Pull request closed without merging**